checker/src/org/checkerframework/checker/lock/LockVisitor.java

```diff
@@ -133,166 +411,116 @@ protected boolean checkOverride(MethodTree overriderTree,
+                    break;
+                default:
+                    expr = tree;
+                    break;
+            }
 
-                            // A GuardedBy annotation on the receiver of the enclosing method was found.
-                            // Now check if the receiver on the enclosing method is the same as the receiver
-                            // on the field/method we are checking preconditions for.
+            Element invokedElement = TreeUtils.elementFromUse(expr);
 
-                            if (guardedByValue != null) {
-                                Node nodeNode = atypeFactory.getNodeForTree(tree);
-                                Node receiverNode = null;
+            AnnotatedTypeMirror atmOfReceiverOrIdentifier = atypeFactory.getReceiverType(tree);
 
-                                if (nodeNode instanceof FieldAccessNode) {
-                                    receiverNode = ((FieldAccessNode) nodeNode).getReceiver();
-                                }
-                                else if (nodeNode instanceof MethodInvocationNode)
-                                {
-                                    MethodAccessNode man = ((MethodInvocationNode) nodeNode).getTarget();
-
-                                    if (!man.getMethod().getModifiers().contains(Modifier.STATIC)) {
-                                        receiverNode = man.getReceiver();
-                                    }
-                                }
+            Node node = atypeFactory.getNodeForTree(tree);
 
-                                if (receiverNode instanceof ExplicitThisLiteralNode ||
-                                    receiverNode instanceof ImplicitThisLiteralNode) {
+            // Is the receiver of the expression being accessed the same as the receiver of the enclosing method?
+            boolean receiverIsThatOfEnclosingMethod = false;
 
-                                    // The receivers match. Add to the preconditions set.
-                                    for (String lockExpression : guardedByValue) {
+            Node receiverNode = null;
 
-                                        if (lockExpression.equals("itself")) {
-                                            // This is critical. That's because right now we know that, since
-                                            // we are dealing with the receiver of the method, "itself" corresponds
-                                            // to "this". However once super.checkPreconditions is called, that
-                                            // knowledge is lost and it will think that "itself" is referring to
-                                            // the variable the precondition we are about to add is attached to.
-                                            lockExpression = "this";
-                                        }
+            if (node instanceof FieldAccessNode || node instanceof MethodAccessNode) {
```

Comment:
I've seen this happen for method calls on implicit `this` receivers. But thanks for pointing this out - `checkAccessOfExpression` should handle dereferences only, not method calls. I added the following at the top of the method to prevent the redundant handling of method calls that was present in `checkAccessOfExpression`:

        if (node instanceof MethodAccessNode) {
            return; // Method calls are not dereferences.
        }

Here is what the implementation looks like now regarding calls to `checkPreconditions` in LockVisitor:

For dereferences (accesses to fields):
`checkPreconditions` is called from `checkAccessOfExpression` when a field of an expression of type `@GuardedBy(...)` is accessed.

For method calls:
`checkPreconditions` is called from `skipReceiverSubtypeCheck` to handle pseudo-assignments of method call receivers and from `commonAssignmentCheck` to handle pseudo-assignments of method parameters. In both cases, `checkPreconditions` is called when the LHS is `@GuardSatisfied` and the RHS is `@GuardedBy`, since this is the scenario in which we must check if the appropriate locks are held. In both cases, there are no dereferences involved, only pseudo-assignments.